File: src/main/java/org/kite9/diagram/visualization/planarization/mapping/ElementMapperImpl.java
package org.kite9.diagram.visualization.planarization.mapping;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import org.apache.commons.math.fraction.BigFraction;
import org.kite9.diagram.adl.Connected;
import org.kite9.diagram.adl.Connection;
import org.kite9.diagram.adl.Container;
import org.kite9.diagram.adl.Diagram;
import org.kite9.diagram.adl.DiagramElement;
import org.kite9.diagram.common.BiDirectional;
import org.kite9.diagram.common.elements.ConnectedVertex;
import org.kite9.diagram.common.elements.MultiCornerVertex;
import org.kite9.diagram.common.elements.PlanarizationEdge;
import org.kite9.diagram.common.elements.Vertex;
import org.kite9.diagram.common.objects.OPair;
import org.kite9.diagram.position.Direction;
import org.kite9.diagram.position.Layout;
import org.kite9.diagram.style.BorderTraversal;
import org.kite9.diagram.visualization.planarization.grid.GridPositioner;
import org.kite9.framework.common.Kite9ProcessingException;
import org.kite9.framework.logging.LogicException;
import org.kite9.framework.serialization.CSSConstants;
import org.kite9.framework.serialization.EnumValue;

public class ElementMapperImpl implements ElementMapper {
	
	private GridPositioner gp;

	public ElementMapperImpl(GridPositioner gp) {
		super();
		this.gp = gp;
	}

	Map<DiagramElement, Vertex> singleVertices = new HashMap<DiagramElement, Vertex>();
	Map<DiagramElement, CornerVertices> cornerVertices = new HashMap<DiagramElement, CornerVertices>();
	Map<BiDirectional<Connected>, PlanarizationEdge> edges = new HashMap<BiDirectional<Connected>, PlanarizationEdge>();
	
	public boolean hasOuterCornerVertices(DiagramElement d) {
		return cornerVertices.containsKey(d);
	}
	
	public CornerVertices getOuterCornerVertices(final DiagramElement c) {
		CornerVertices v = cornerVertices.get(c);
		if (v == null) {
			if (hasParentGridLayout(c)) {
				CornerVertices parentCV = getOuterCornerVertices((Container)c.getParent());
				OPair<BigFraction> xspan = gp.getGridXPosition(c);
				OPair<BigFraction> yspan = gp.getGridYPosition(c);
				v = new SubwindowCornerVertices(c, xspan, yspan, parentCV);
				cornerVertices.put(c, v);
			} else {
				v = new IndependentCornerVertices(c);
				cornerVertices.put(c, v);
			}
		}

		return v;
	}

	private boolean hasParentGridLayout(DiagramElement c) {
		DiagramElement parent = c.getParent();
		if ((parent != null) && (parent instanceof Container)) {
			return ((Container)parent).getLayout()==Layout.GRID;
		}
		
		return false;
	}

	public PlanarizationEdge getEdge(Connected from, Vertex vfrom, Connected to, Vertex vto,
			BiDirectional<Connected> element) {
		PlanarizationEdge e = edges.get(element);

		Direction dd = element.getDrawDirectionFrom(from);

		if (e == null) {
			if (element instanceof Connection) {
				e = new ConnectionEdge(vfrom, vto, (Connection) element, dd);
			} else if (element instanceof GeneratedLayoutElement) {
				e = new ContainerLayoutEdge(vfrom, vto, dd, from, to);
			} else {
				throw new LogicException("Unknown BiDirectional type: "+element);
			}

			if (element != null) {
				edges.put(element, e);
			}

			return e;
		} else {
			Vertex oldFrom = e.getFrom();
			oldFrom.removeEdge(e);
			vfrom.addEdge(e);
			e.setFrom(vfrom);

			Vertex oldTo = e.getTo();
			oldTo.removeEdge(e);
			vto.addEdge(e);
			e.setTo(vto);

			e.setDrawDirectionFrom(dd, vfrom);
			return e;
		}
	}

	public Vertex getPlanarizationVertex(DiagramElement c) {
		Vertex v = singleVertices.get(c);
		if (v == null) {
			
			if (c instanceof Connected) {
				v = new ConnectedVertex(c.getID(), (Connected) c);
				singleVertices.put(c, v);
			} else {
				throw new Kite9ProcessingException("Not sure how to create vertex for "+c);
			}
		}

		return v;
	}

	public boolean hasCreated(Connection element) {
		return edges.get(element) != null;
	}

	/**
	 * Debug only - very slow
	 */
	@Override
	public Collection<Vertex> allVertices() {
		Collection<Vertex> out = new ArrayList<Vertex>(singleVertices.values());
		for (CornerVertices cv : cornerVertices.values()) {
			for (MultiCornerVertex vertex : cv.getVerticesAtThisLevel()) {
				out.add(vertex);
			}
		}
		
		return out;
	}

	private Map<DiagramElement, Integer> containerDepths = new HashMap<DiagramElement, Integer>(100);

	@Override
	public int getContainerDepth(DiagramElement c) {
		DiagramElement parent = c.getParent();
		if (parent==null) {
			return 0;
		} else {
			Integer depth = containerDepths.get(c);
			if (depth != null) {
				return depth;
			} else {
				// we don't count nested grids, because these containers are parts of the whole.
				
				Layout l = ((Container) parent).getLayout();
				depth = getContainerDepth(parent) + ((l==Layout.GRID) ? 0 : 1);
				containerDepths.put(c, depth);
			}
			return depth;
		}
	}
	
	public boolean requiresPlanarizationCornerVertices(DiagramElement c) {
		if (c instanceof Diagram) {
			return true;
		}
		// does anything inside it have connections?
		if (c instanceof Container) {
			for (DiagramElement de : ((Container) c).getContents()) {
				if (hasNestedConnections(de)) {
					return true;
				}
			}
			
			// are connections allowed to pass through it?
			boolean canTraverse = isElementTraversible(c);
			if (canTraverse && hasNestedConnections(c)) {
				return true;
			}
		}
		
		// is it embedded in a grid?  If yes, use corners
		Layout l = c.getParent() == null ? null : ((Container) c.getParent()).getLayout();
		return (l == Layout.GRID);
	}

	private boolean isElementTraversible(DiagramElement c) {
		return isElementTraversible(c, CSSConstants.TRAVERSAL_BOTTOM_PROPERTY) ||
				isElementTraversible(c, CSSConstants.TRAVERSAL_LEFT_PROPERTY) ||
				isElementTraversible(c, CSSConstants.TRAVERSAL_RIGHT_PROPERTY) ||
				isElementTraversible(c, CSSConstants.TRAVERSAL_TOP_PROPERTY);
	}

	private boolean isElementTraversible(DiagramElement c, String p) {
		EnumValue v = (EnumValue) c.getCSSStyleProperty(p);
		BorderTraversal bt = (BorderTraversal) v.getTheValue();
		return (bt == BorderTraversal.ALWAYS);
	}

	Map<DiagramElement, Boolean> hasConnections = new HashMap<>();
	
	public boolean hasNestedConnections(DiagramElement c) {
		if (hasConnections.containsKey(c)) {
			return hasConnections.get(c);
		} 
		
		boolean has = false;
		
		if (c instanceof Connected) {
			has = ((Connected)c).getLinks().size() > 0;
		}
		
		if ((has == false) && (c instanceof Container)) {
			for (DiagramElement de : ((Container)c).getContents()) {
				if (hasNestedConnections(de)) {
					has = true;
					break;
				}
			}
		}
		
		hasConnections.put(c, has);
		return has;
	}
	
}

File: k9layout/src/main/kotlin/org/kite9/diagram/common/elements/mapping/ElementMapperImpl.kt
package org.kite9.diagram.common.elements.mapping

import org.kite9.diagram.common.BiDirectional
import org.kite9.diagram.common.elements.edge.PlanarizationEdge
import org.kite9.diagram.common.elements.grid.GridPositioner
import org.kite9.diagram.common.elements.vertex.ConnectedRectangularVertex
import org.kite9.diagram.common.elements.vertex.Vertex
import org.kite9.diagram.common.fraction.LongFraction
import org.kite9.diagram.logging.LogicException
import org.kite9.diagram.model.*
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.model.position.HPos
import org.kite9.diagram.model.position.Layout
import org.kite9.diagram.model.position.VPos
import org.kite9.diagram.model.style.BorderTraversal

class ElementMapperImpl(private val gp: GridPositioner) : ElementMapper {

    private var singleVertices: MutableMap<DiagramElement, Vertex> = HashMap()
    private var cornerVertices: MutableMap<DiagramElement, CornerVertices> = HashMap()
    private var baseGrids: MutableMap<DiagramElement?, BaseGridCornerVertices> = HashMap()
    private var edges: MutableMap<BiDirectional<Connected>, PlanarizationEdge> = HashMap()
    private var hasConnections: MutableMap<DiagramElement?, Boolean> = HashMap()

    override fun hasOuterCornerVertices(d: DiagramElement): Boolean {
        return cornerVertices.containsKey(d)
    }

    override fun getOuterCornerVertices(c: DiagramElement): CornerVertices {
        var v = cornerVertices[c]
        if (v == null) {
            if (isEmbeddedWithinGrid(c)) {
                val parentCV = getBaseGridCornerVertices(c.getParent() as Container?)
                v = createSubGridCornerVertices(c, parentCV)
            } else {
                v = IndependentCornerVertices(c, c.getDepth())
                cornerVertices[c] = v
            }
        }
        return v
    }

    private fun createSubGridCornerVertices(
        c: DiagramElement,
        parentCV: BaseGridCornerVertices
    ): SubGridCornerVertices {
        val xspan = (c as Rectangular).getRenderingInformation().gridXPosition()
        val yspan = c.getRenderingInformation().gridYPosition()
        val v = SubGridCornerVertices(c, xspan!!, yspan!!, parentCV, c.getDepth())
        cornerVertices[c] = v
        return v
    }

    private fun getBaseGridCornerVertices(c: Container?): BaseGridCornerVertices {
        var bgcv = baseGrids[c]
        if (bgcv == null) {
            bgcv = BaseGridCornerVertices(c!!, c!!.getDepth() + 1)
            baseGrids[c] = bgcv
        }
        return bgcv
    }

    private fun isEmbeddedWithinGrid(c: DiagramElement): Boolean {
        val parent = c.getParent()
        if (c is Label) {
            return false
        }
        return if (parent != null && parent is Container) {
            parent.getLayout() == Layout.GRID
        } else false
    }

    override fun getEdge(
        from: Connected, vfrom: Vertex, to: Connected, vto: Vertex,
        element: BiDirectional<Connected>
    ): PlanarizationEdge? {
        var e = edges[element]
        val dd = element.getDrawDirectionFrom(from)
        return if (e == null) {
            e = if (element is Connection) {
                ConnectionEdge(vfrom, vto, element, dd)
            } else if (element is GeneratedLayoutBiDirectional) {
                ContainerLayoutEdge(vfrom, vto, dd!!, from, to)
            } else {
                throw LogicException("Unknown BiDirectional type: $element")
            }
            if (element != null) {
                edges[element] = e
            }
            e
        } else {
            val oldFrom = e.getFrom()
            oldFrom.removeEdge(e)
            vfrom.addEdge(e)
            e.setFrom(vfrom)
            val oldTo = e.getTo()
            oldTo.removeEdge(e)
            vto.addEdge(e)
            e.setTo(vto)
            e.setDrawDirectionFrom(dd, vfrom)
            e
        }
    }

    override fun getPlanarizationVertex(de: DiagramElement): Vertex {
        var v = singleVertices[de]
        if (v == null) {
            if (de is ConnectedRectangular) {
                v = ConnectedRectangularVertex(de.getID(), de)
                singleVertices[de] = v
            }else if (de is Port) {
                val c1 = de.getContainer()!!
                val cv = getOuterCornerVertices(c1)
                val (fracX, fracY) = buildFractions(c1, de)
                v = cv.createVertex(fracX, fracY,
                    HPos.getFromDirection(de.getPortDirection()), VPos.getFromDirection(de.getPortDirection()),
                    c1, de)

                singleVertices[de] = v
            } else {
                throw LogicException("Not sure how to create vertex for $de")
            }
        }
        return v
    }

    private fun buildFractions(c1: Container, de: Port): Pair<LongFraction, LongFraction> {
        val portsOnSide = c1.getContents()
            .filterIsInstance<Port>()
            .filter { it.getPortDirection() == de.getPortDirection() }
            .sortedBy { it.getPortPosition() }

        val num = portsOnSide.indexOf(de) + 1
        val denom = portsOnSide.size + 1
        val fraction = LongFraction(num.toLong(), denom.toLong())

        return when (de.getPortDirection()) {
            Direction.UP -> Pair(fraction, LongFraction.ZERO)
            Direction.DOWN -> Pair(fraction, LongFraction.ONE)
            Direction.LEFT -> Pair(LongFraction.ZERO, fraction)
            Direction.RIGHT -> Pair(LongFraction.ONE, fraction)
        }
    }

    /**
     * Debug only - very slow
     */
    override fun allVertices(): Collection<Vertex> {
        val out: MutableCollection<Vertex> = ArrayList(singleVertices.values)
        for (cv in cornerVertices.values) {
            for (vertex in cv.getVerticesAtThisLevel()) {
                out.add(vertex)
            }
        }
        return out
    }

    override fun requiresPlanarizationCornerVertices(c: DiagramElement): Boolean {
        if (c is Diagram) {
            return true
        }
        // does anything inside it have connections?
        if (c is Container) {
            for (de in c.getContents()) {
                if (hasNestedConnections(de)) {
                    return true
                }
            }

            // are connections allowed to pass through it?
            val canTraverse = isElementTraversible(c)
            if (canTraverse && hasNestedConnections(c)) {
                return true
            }
        }

        // is it embedded in a grid?  If yes, use corners
        if (c is ConnectedRectangular) {
            val l = if (c.getParent() == null) null else (c.getParent() as Container?)!!.getLayout()
            return l == Layout.GRID
        }

        return false
    }

    private fun isElementTraversible(c: DiagramElement): Boolean {
        return isElementTraversible(c, Direction.UP) ||
                isElementTraversible(c, Direction.DOWN) ||
                isElementTraversible(c, Direction.LEFT) ||
                isElementTraversible(c, Direction.RIGHT)
    }

    private fun isElementTraversible(c: DiagramElement, d: Direction): Boolean {
        return if (c is Container) {
            c.getTraversalRule(d) == BorderTraversal.ALWAYS
        } else false
    }

    fun hasNestedConnections(c: DiagramElement?): Boolean {
        if (hasConnections.containsKey(c)) {
            return hasConnections[c]!!
        }
        var has = false
        if (c is Connected) {
            has = c.getLinks().size > 0
        }
        if (has == false && c is Container) {
            for (de in c.getContents()) {
                if (hasNestedConnections(de)) {
                    has = true
                    break
                }
            }
        }
        hasConnections[c] = has
        return has
    }

    override fun getGridPositioner(): GridPositioner {
        return gp
    }
}